File: python/popart._internal.ir/bindings/transforms/decomposesum.cpp
// Copyright (c) 2021 Graphcore Ltd. All rights reserved.
#include "bindings/transforms/decomposesum.hpp"
#include "bindings/transforms/transform.hpp"

#include <array>
#include <initializer_list>
#include <pybind11/pybind11.h>
#include <popart/transforms/decomposesum.hpp>

#include <popart/graph.hpp> // IWYU pragma: keep

namespace py = pybind11;

namespace popart {
class Transform;

namespace _internal {
namespace ir {
namespace transforms {
void bindDecomposeSum(py::module &m) {

  py::class_<DecomposeSum, Transform, PyTransform<DecomposeSum>>(m,
                                                                 "DecomposeSum")
      .def(py::init<>())
      .def("apply", &DecomposeSum::apply)
      .def("getId", &DecomposeSum::getId)
      .def("getName", &DecomposeSum::getName);
}

} // namespace transforms
} // namespace ir
} // namespace _internal
} // namespace popart

File: python/popxl/python_files/transforms/decompose_sum.py
# Copyright (c) 2021 Graphcore Ltd. All rights reserved.
"""Decompose Gradients Sum transform."""
import popart._internal.ir as _ir
from popxl.graph import Graph


def decompose_sum(graph: Graph) -> Graph:
    """
    Transform the input Graph by decomposing Sum operations with >2 inputs into a liveness optimal tree of additions.

    Args:
        graph (Graph):
            The graph to decompose sum.

    Returns:
        decomposed_graph (:py:class:`popxl.Graph`): decomposed gradient sum graph.
    """
    transform = _ir.transforms.DecomposeSum()
    transform.apply(graph._pb_graph)
    return graph
